### ISSUE TITLE
RFC: Boilerplate is abandoned

### ISSUE THREAD

**Issue by saibotsivad** [ngbp/ngbp#347]
This is not meant as a criticism, it is an open request for comments from the developers and the community:
- It looks like the [last commit]() was a while ago (25 days currently).
- There are [52 open pull requests](https://github.com/ngbp/ngbp/pulls), some of them very old.
- Even trivial-but-important PRs, like [this one](https://github.com/ngbp/ngbp/pull/337), have been sitting for a month and a half.

On the one hand, I could fork boilerplate myself and merge a bunch of pull requests, but a few months down the road it would be sitting in the same position: I would have lost interest, and the code would get stagnant. I don't want to maintain even a cool project like this forever!

One potential solution I would suggest, that Rod Vagg uses in [levelup](https://github.com/rvagg/node-levelup) is making people contributors. (See his [working manifesto](https://github.com/rvagg/node-levelup/blob/master/CONTRIBUTING.md) as an example.) In practice what this means is that anyone who contributes "reasonably significant" code is nearly automatically made a contributor.

This solves the problem of a lack of people with admin power being able to merge pull requests.

Comments? Critiques? Suggestions?

**Comment by danielbsig** [comment]
Agree. I'm using ngbp myself, and it would be good to know that this project isn't dead.

**Comment by joshdmiller** [comment]
I appreciate the thoughtful feedback. The project is not dead, but we need both more contributors and higher quality code contributions.

While most people assume that I'm doing this solo, ngbp in fact has 4 contributors with commit access and I am always willing to add more if there's something one wants to accomplish - present company included. :-)

You mentioned that there are 52 open PRs. While true, most of those cannot be merged as they do not follow the commit guidelines or are generally of poor quality. For example, PR #337 to which you linked does not contain a valid commit message. Many are even duplicates, and some are even open only as ideas. There is a communication problem here, but not a code problem (for the most part).

All that said, the Grunt script is already a bit unwieldy, so we probably don't want to entertain too many "enhancements" to ngbp. We should definitely commit to keeping things updated, fixing bugs, replacing tools as necessary, etc., but adding large new features (like Jade support) is probably not a good idea. These can be evaluated on an individual basis, however.

The first thing that needs to be done is to clean up all the open issues and PRs. Most of them need be closed outright as I have done a very poor job in communicating thus far. Second, we should identify what more we wish to accomplish and see who wants to contribute those features.

That's a discussion I'd love to have.

**Comment by TehShrike** [comment]
Personally I don't think features are particularly needed.  I think it is very reasonable to just focus on keeping ngbp up to date with current versions of AngularJS and the other included modules.

---

Maybe this should be a different issue, but I feel picky contribution requirements cause more harm than good in projects like this.  Better to merge in a commit with a sub-optimal message than to leave the code unfixed.

**Comment by saibotsivad** [comment]
@joshdmiller I'm glad you're still around! Sorry for the length of the text below. :)

I saw in the ngbp organization that there were four other contributors, but based on the lack of activity it appears that the four of you don't have enough free time to devote to maintaining the project.

For example, on the first page of pull requests, only 8/25 seem to have comments from contributors. Of the ones that are commented on by contributors, a number of them seem to indicate that the pull request is ready to be merged, but they have been sitting open for months.

If this is a case of people not making good pull requests (always a possibility) than there would be more comments on the pull requests to indicate what should be changed, ideally a link to some "code style guide" or perhaps a "guiding philosophy" document. (Semi-relatedly, the links in the [wiki](https://github.com/ngbp/ngbp/wiki) go to the old repo, so none of those documents are reachable.)

Again, this isn't meant as a criticism, but due to lack of activity it certainly seems that this project has been abandoned.

Correct me if I'm wrong, but based on you moving the boilerplate from your personal repo to the ngbp, it appears that you intended this project to become less personal and more community driven?

If that's the case, I think that a reasonable solution to keeping ngbp updated and alive is threefold:

**1. Cast a vision**

For other people to volunteer their time, you'll want to say "_this_ is the goal". This might be as simple as writing up a "design philosophy" page in the wiki.

**2. Lower the bar for adding contributors.**

This seems to be the approach of [Rod Vagg](https://github.com/rvagg) (of levelup fame). In practice what this looks like is that anyone who contributes some arbitrarily long/reasonable code is automatically added as a contributor. (They can remove themselves if they don't want to be.)

The advantage of this is:
- It encourages community involvement. Giving people responsibility makes them want to be more involved.
- Less personal work required. Other people can address issues and merge pull requests, meaning that you don't have to do as much work.

**3. Lower the bar for pull requests.**

(I don't know if this is actually necessary, if you have enough active contributors.)

If pull requests don't have comments in the commits, or other such things, maybe they can still be merged if they are not important (e.g. slight wording change, fix links, etc.).

---

All this to say: I really appreciate all the work that's gone into this project, it's been an indispensable help in learning grunt and angular! I'd love to help mature this project and make it more community-driven.

**Comment by saibotsivad** [comment]
(This is just a reminder to read Rod Vagg's [OPEN Open Source "Manifesto"](https://github.com/rvagg/node-levelup/blob/master/CONTRIBUTING.md).)

**Comment by joshdmiller** [comment]
Hey @saibotsivad - I appreciate the feedback and I don't take any of it personally.

First, I moved ngbp from my personal account to an org with the express purpose of having this more community-driven - and also setting the expectation that I would not be personally responsible for making or merging all changes. So you're right here.

Also, the bar for adding contributors is as low as is reasonable. I have never turned anyone away who has asked to help, but there is a level of trust that all committers will adhere to the core vision and to Git best practices. If I missed someone you feel should be a contributor, let's add her. But I don't think Rod's approach would have resulted in any more contributors in this case - most PRs are very small and from a diverse assortment different people.

For the record, there already _is_ [such a commit guide](https://github.com/ngbp/ngbp/blob/v0.3.2-release/CONTRIBUTING.md) and, in true GitHub fashion, a link to it automatically appears at the top of the page whenever anyone creates a new issue or PR. This has only been in place for a few months, but it is there now and it is reasonable to expect contributors to follow it.

Speaking of which, the guidelines are very straightforward and the same as those followed by AngularJS, AngularUI, Karma, and a number of other projects. The core of this philosophy is that each commit represents one logical, atomic change, so having a proper commit message means that out changelog _is_ our release notes. This saves time and effort. We should definitely be guiding potential contributors in this direction (as I mentioned in my previous comment, there is a communication problem), but we should not be relaxing these requirements just for the sake of getting more code in. That's actually counterproductive.

I agree with adding a vision, particularly considering how much ngbp has evolved over the past few years. The core of that is still [in the readme](https://github.com/ngbp/ngbp#philosophy), but adding a little more on scope may be helpful. I'll put some thought into that.

In the meantime, can you put together a list of pull requests you'd like to see merged that you feel have been neglected?  Listing them in a comment on this thread would be just fine. Then we can take a look and we draw a map from where we are to where we want to be.

**Comment by saibotsivad** [comment]
I admittedly haven't thought through Rod's approach in depth, so even though it seems to me like a good approach to projects that are heavily community driven, I'm not going to try pushing it very hard.

When you speak about "core vision", are you talking about the "philosophy" section in the README file? There's some stuff linked to in the wiki, but the links are broken.

Things being self documenting is obviously awesome, so I do like the idea of the commits being the changelog. And I agree with you that, as a developer, I don't want to relax requirements just to push more things through. I've been thinking through that issue, trying to think if a particular Github work flow implementation would allow merging of simple things [like this](https://github.com/ngbp/ngbp/pull/337) comma fix without the "strict" commit message requirements. I've not come up with a good solution, so I think I'm in agreement with you there.

Probably the thing to do here is be more vocal about pull requests not fitting the commit guide, and to be better at labeling pull requests/issues, and getting things merged.

I'm interested in becoming a contributor. The company I work for uses several instances of ngbp for reasonably large projects, so internally we've been talking through some of the issues and best practices.

My goals as a contributor would probably be (roughly in this order):
- Go through the pull requests, adding labels and closing where appropriate.
- There's a bunch of places in the docs that still point to your old repo, and some other stuff like that that should be updated.
- Locally merge everything, make sure it all works and looks nice and tidy, then I'll ping you and we can figure out how to merge and make a new version, etc.

**Comment by benjaminsuch** [comment]
I would also like to contribute. But as far as I can see, there isn't a goal defined in any of the .md's?

**Comment by joshdmiller** [comment]
@saibotsivad Those are great goals and I stand behind them.

I've invited you and @benjaminsuch to be collaborators on ngbp - welcome aboard!

I do ask that we have a little chat to connect before we push anything live so we can ensure we're on the same page. However, it would be **awesome** if you guys went through the PRs and found what's good, what's not, and decided which ones you want to merge. If you have new ideas, definitely create issues for those; I can't commit too much time to ngbp right now, but I'll definitely pay attention to issue filings and chime in if I have a _problem_ with something - otherwise, community rules. Yay!

PS: You mentioned the wiki a couple times, @saibotsivad, and I had no idea what you were talking about, so I didn't pay too much attention to it. Last night, I took a look and it seems I created that page about year ago...but I don't remember doing so. :confused: Anyway, I removed that table of contents to avoid confusion - all documentation is in the READMEs right now.
